future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,540 @@
+	var sigTxs [][]byte
+	var txids []string
+	for _, txAndSigner := range txs {
+		txGroup := []types.Transaction{txAndSigner.Txn}
+		indexesToSign := []int{0}
+		sigStx, txid, err := txAndSigner.Signer(txGroup, indexesToSign)
+		if err != nil {
+			return nil, err
+		}
+
+		sigTxs = append(sigTxs, sigStx[0])
+		txids = append(txids, txid[0])
+	}
+
+	atc.signedTxs = sigTxs
+	atc.txids = txids
+	atc.status = SIGNED
+	return sigTxs, nil
+}
+
+/**
+* Send the transaction group to the network, but don't wait for it to be committed to a block. An
+* error will be thrown if submission fails.
+*
+* The composer's status must be SUBMITTED or lower before calling this method. If submission is
+* successful, this composer's status will update to SUBMITTED.
+*
+* Note: a group can only be submitted again if it fails.
+*
+* @returns A promise that, upon success, resolves to a list of TxIDs of the submitted transactions.
+ */
+func (atc *AtomicTransactionComposer) Submit(client *algod.Client) ([]string, error) {
+	if atc.status > SUBMITTED {
+		return nil, errors.New("status must be SUBMITTED or lower in order to call Submit()")
+	}
+
+	if atc.status == SUBMITTED {
+		return atc.txids, nil
+	}
+
+	stxs, err := atc.GatherSignatures()
+	if err != nil {
+		return nil, err
+	}
+
+	var serializedStxs []byte
+	for _, stx := range stxs {
+		serializedStxs = append(serializedStxs, stx...)
+	}
+
+	_, err = client.SendRawTransaction(serializedStxs).Do(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	return atc.txids, nil
+}
+
+/**
+* Send the transaction group to the network and wait until it's committed to a block. An error
+* will be thrown if submission or execution fails.
+*
+* The composer's status must be SUBMITTED or lower before calling this method, since execution is
+* only allowed once. If submission is successful, this composer's status will update to SUBMITTED.
+* If the execution is also successful, this composer's status will update to COMMITTED.
+*
+* Note: a group can only be submitted again if it fails.
+*
+* @returns A promise that, upon success, resolves to an object containing the confirmed round for
+*   this transaction, the txIDs of the submitted transactions, and an array of results containing
+*   one element for each method call transaction in this group. If a method has no return value
+*   (void), then the method results array will contain null for that method's return value.
+ */
+func (atc *AtomicTransactionComposer) Execute(client *algod.Client) (int, []string, []ABIValue, error) {
+	if atc.status > SUBMITTED {
+		return 0, nil, nil, errors.New("status is already committed")
+	}
+
+	_, err := atc.Submit(client)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	txinfo, err := WaitForConfirmation(client, atc.txids[0], 0, context.Background())
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	var returnValues []ABIValue
+	for i, txid := range atc.txids {
+		if atc.transactions[i].Txn.Type != types.ApplicationCallTx {
```

Comment:
I *think* we should be safe and refer back to our `Method` map and check if the `i` is in the map. I believe `ApplicationCallTx` could be sent as a regular transaction (as part of a `TransactionWithSigner`) because we allow `appl` calls as method arguments.